File: Chatbots/Python/Source/tokenise.py
from __future__ import annotations

from typing import List, Set
from .config import ParserConfig


def _emit_token(buffer: List[str], stopwords: Set[str], config: ParserConfig, output: List[str]) -> None:
    """Apply length/stopword rules and append to output if accepted."""
    if not buffer:
        return
    token = "".join(buffer)
    buffer.clear()

    if token in stopwords and config.remove_stopwords:
        return

    if token.isdigit() and config.keep_digits:
        output.append(token)
        return

    if len(token) >= config.min_token_length:
        output.append(token)


def tokenise(text: str, stopwords: Set[str], config: ParserConfig) -> List[str]:
    """
    Split on non-alphanumeric boundaries and return filtered tokens.
    Assumes text was already normalised (e.g., lowercased) upstream.
    """
    if not text:
        return []

    output: List[str] = []
    buffer: List[str] = []

    for character in text:
        if character.isalnum():
            buffer.append(character)
        else:
            _emit_token(buffer, stopwords, config, output)

    _emit_token(buffer, stopwords, config, output)
    return output


def tokenise_to_set(text: str, stopwords: Set[str], config: ParserConfig) -> Set[str]:
    """Tokenise and return a set for fast overlap operations."""
    return set(tokenise(text, stopwords, config))

File: Chatbots/Rust/Source/src/tokenise.rs
use std::collections::{HashSet, hash_map::RandomState};

use crate::config::ParserConfig;

/// Determine whether a token consists only of digits.
fn token_is_numeric(token_text: &str) -> bool {
    token_text.chars().all(|character| character.is_ascii_digit())
}

/// Split on non-alphanumeric characters, keep only tokens that pass length rules,
/// and remove stopwords if configured. Returns tokens in the original order.
pub fn tokenise(
    input_text: &str,
    stopword_set: &HashSet<String>,
    parser_config: &ParserConfig,
) -> Vec<String> {
    let mut tokens: Vec<String> = Vec::new();
    let mut current_token = String::new();

    // Build tokens using Unicode-aware classification.
    for character in input_text.chars() {
        if character.is_alphanumeric() {
            current_token.push(character.to_ascii_lowercase());
        } else {
            if !current_token.is_empty() {
                tokens.push(current_token.clone());
                current_token.clear();
            }
        }
    }
    if !current_token.is_empty() {
        tokens.push(current_token.clone());
        current_token.clear();
    }

    // Apply length and stopword rules
    let mut filtered_tokens: Vec<String> = Vec::with_capacity(tokens.len());
    for token_text in tokens.into_iter() {
        let token_length = token_text.chars().count();
        let is_numeric = token_is_numeric(&token_text);
        let meets_length_rule =
            token_length >= parser_config.min_token_length || is_numeric;

        if !meets_length_rule {
            continue;
        }
        if parser_config.remove_stopwords && stopword_set.contains(&token_text) {
            continue;
        }
        filtered_tokens.push(token_text);
    }

    filtered_tokens
}

/// As above, but return a set of unique tokens (order not preserved).
pub fn tokenise_to_set(
    input_text: &str,
    stopword_set: &HashSet<String>,
    parser_config: &ParserConfig,
) -> HashSet<String> {
    let sequence_tokens = tokenise(input_text, stopword_set, parser_config);
    sequence_tokens.into_iter().collect::<HashSet<String, RandomState>>()
}
